File: Bookcase core/src/test/java/cz/vhromada/book/validator/CategoryValidatorTest.java
package cz.vhromada.book.validator;

import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.assertj.core.api.SoftAssertions.assertSoftly;
import static org.mockito.Mockito.verifyZeroInteractions;

import java.util.Collections;

import cz.vhromada.book.entity.Category;
import cz.vhromada.book.utils.CategoryUtils;
import cz.vhromada.common.Movable;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.test.validator.MovableValidatorTest;
import cz.vhromada.common.validator.MovableValidator;
import cz.vhromada.common.validator.ValidationType;
import cz.vhromada.result.Event;
import cz.vhromada.result.Result;
import cz.vhromada.result.Severity;
import cz.vhromada.result.Status;

import org.junit.jupiter.api.Test;

/**
 * A class represents test for class {@link CategoryValidator}.
 *
 * @author Vladimir Hromada
 */
class CategoryValidatorTest extends MovableValidatorTest<Category, cz.vhromada.book.domain.Category> {

    /**
     * Test method for {@link CategoryValidator#CategoryValidator(MovableService)} with null service for categories.
     */
    @Test
    void constructor_NullCategoryService() {
        assertThatThrownBy(() -> new CategoryValidator(null)).isInstanceOf(IllegalArgumentException.class);
    }

    /**
     * Test method for {@link CategoryValidator#validate(Movable, ValidationType...)} with {@link ValidationType#DEEP} with data with null name.
     */
    @Test
    void validate_Deep_NullName() {
        final Category category = getValidatingData(1);
        category.setName(null);

        final Result<Void> result = getMovableValidator().validate(category, ValidationType.DEEP);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents())
                .isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "CATEGORY_NAME_NULL", "Name mustn't be null.")));
        });

        verifyZeroInteractions(getMovableService());
    }

    /**
     * Test method for {@link CategoryValidator#validate(Movable, ValidationType...)} with {@link ValidationType#DEEP} with data with empty name.
     */
    @Test
    void validate_Deep_EmptyName() {
        final Category category = getValidatingData(1);
        category.setName("");

        final Result<Void> result = getMovableValidator().validate(category, ValidationType.DEEP);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents())
                .isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "CATEGORY_NAME_EMPTY", "Name mustn't be empty string.")));
        });

        verifyZeroInteractions(getMovableService());
    }

    @Override
    protected MovableValidator<Category> getMovableValidator() {
        return new CategoryValidator(getMovableService());
    }

    @Override
    protected Category getValidatingData(final Integer id) {
        return CategoryUtils.newCategory(id);
    }

    @Override
    protected Category getValidatingData(final Integer id, final Integer position) {
        final Category category = CategoryUtils.newCategory(id);
        category.setPosition(position);

        return category;
    }

    @Override
    protected cz.vhromada.book.domain.Category getRepositoryData(final Category validatingData) {
        return CategoryUtils.newCategoryDomain(validatingData.getId());
    }

    @Override
    protected cz.vhromada.book.domain.Category getItem1() {
        return CategoryUtils.newCategoryDomain(1);
    }

    @Override
    protected cz.vhromada.book.domain.Category getItem2() {
        return CategoryUtils.newCategoryDomain(2);
    }

    @Override
    protected String getName() {
        return "Category";
    }

}

File: bookcase-core/src/test/kotlin/cz/vhromada/bookcase/validator/CategoryValidatorTest.kt
package cz.vhromada.bookcase.validator

import com.nhaarman.mockitokotlin2.verifyZeroInteractions
import cz.vhromada.bookcase.entity.Category
import cz.vhromada.bookcase.utils.CategoryUtils
import cz.vhromada.common.result.Event
import cz.vhromada.common.result.Severity
import cz.vhromada.common.result.Status
import cz.vhromada.common.test.validator.MovableValidatorTest
import cz.vhromada.common.validator.MovableValidator
import cz.vhromada.common.validator.ValidationType
import org.assertj.core.api.SoftAssertions.assertSoftly
import org.junit.jupiter.api.Test

/**
 * A class represents test for class [CategoryValidator].
 *
 * @author Vladimir Hromada
 */
class CategoryValidatorTest : MovableValidatorTest<Category, cz.vhromada.bookcase.domain.Category>() {

    /**
     * Test method for [CategoryValidator.validate] with [ValidationType.DEEP] with data with null name.
     */
    @Test
    fun validateDeepNullName() {
        val category = getValidatingData(1)
                .copy(name = null)

        val result = getValidator().validate(category, ValidationType.DEEP)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "CATEGORY_NAME_NULL", "Name mustn't be null.")))
        }

        verifyZeroInteractions(service)
    }

    /**
     * Test method for [CategoryValidator.validate] with [ValidationType.DEEP] with data with empty name.
     */
    @Test
    fun validateDeepEmptyName() {
        val category = getValidatingData(1)
                .copy(name = "")

        val result = getValidator().validate(category, ValidationType.DEEP)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "CATEGORY_NAME_EMPTY", "Name mustn't be empty string.")))
        }

        verifyZeroInteractions(service)
    }

    override fun getValidator(): MovableValidator<Category> {
        return CategoryValidator(service)
    }

    override fun getValidatingData(id: Int?): Category {
        return CategoryUtils.newCategory(id)
    }

    override fun getValidatingData(id: Int?, position: Int?): Category {
        return CategoryUtils.newCategory(id)
                .copy(position = position)
    }

    override fun getRepositoryData(validatingData: Category): cz.vhromada.bookcase.domain.Category {
        return CategoryUtils.newCategoryDomain(validatingData.id)
    }

    override fun getItem1(): cz.vhromada.bookcase.domain.Category {
        return CategoryUtils.newCategoryDomain(1)
    }

    override fun getItem2(): cz.vhromada.bookcase.domain.Category {
        return CategoryUtils.newCategoryDomain(2)
    }

    override fun getName(): String {
        return "Category"
    }

}
